server/web.js

```diff
@@ -314,3 +381,427 @@ function getPluginVersion(callback) {
+    var indexInCurrentChunkInMerged = 0;
+
+    diff.forEach(function(part) {
+
+      if (part.added) {
+
+        var currentChunk = mergedDiff[currentChunkInMerged];
+
+        // Skip through the already removed chunks
+        // This preserves order if I remove something,
+        //   and then add something later in the same place
+        while (currentChunk && currentChunk.removed) {
+          currentChunkInMerged += 1;
+          currentChunk = mergedDiff[currentChunkInMerged];
+        }
+
+        if (!currentChunk) {
+          // The added part is at the very end of the file
+          mergedDiff.push(part);
+
+          // Our indexes are at the end of the last part
+          currentChunkInMerged = mergedDiff.length;
+          indexInCurrentChunkInMerged = part.value.length;
+
+        } else {
+          // Split up this chunk into previous and next
+          var prevChunk = JSON.parse(JSON.stringify(currentChunk));
+          prevChunk.value = prevChunk.value.substring(0, indexInCurrentChunkInMerged);
+          var nextChunk = JSON.parse(JSON.stringify(currentChunk));
+          nextChunk.value = nextChunk.value.substring(indexInCurrentChunkInMerged);
+
+          // Delete the current chunk and replace it with prev, part, and next
+          mergedDiff.splice(currentChunkInMerged, 1, prevChunk, part, nextChunk);
+
+          // Now, we start at the beginning of nextChunk
+          currentChunkInMerged += 2;
+          indexInCurrentChunkInMerged = 0;
+        }    
+        
+      } else if (part.removed) {
+
+        var currentChunk = mergedDiff[currentChunkInMerged];
+
+        // Skip through the already removed chunks
+        while (currentChunk && currentChunk.removed) {
+          currentChunkInMerged += 1;
+          currentChunk = mergedDiff[currentChunkInMerged];
+        }
+
+        if (indexInCurrentChunkInMerged + part.value.length < currentChunk.value.length) {
+          // The remove is within a single chunk
+          var prevChunk = JSON.parse(JSON.stringify(currentChunk));
+          prevChunk.value = prevChunk.value.substring(0, indexInCurrentChunkInMerged);
+          var deletedChunk = JSON.parse(JSON.stringify(currentChunk));
+          deletedChunk.value = deletedChunk.value.substring(indexInCurrentChunkInMerged, indexInCurrentChunkInMerged + part.value.length);
+          deletedChunk.removed = true;
+          deletedChunk.added = false;
+          var nextChunk = JSON.parse(JSON.stringify(currentChunk));
+          nextChunk.value = nextChunk.value.substring(indexInCurrentChunkInMerged + part.value.length);
+
+          // Delete the current chunk and replace it
+          mergedDiff.splice(currentChunkInMerged, 1, prevChunk, deletedChunk, nextChunk);
+
+          // Starting at the beginning of nextChunk
+          currentChunkInMerged += 2;
+          indexInCurrentChunkInMerged = 0;
+
+        } else {
+          // The remove goes over multiple chunks
```

Comment:
Can deferring some work for the next time around the loop reduce some of the if/else and duplication?

Reply:
Remaining challenge, now that you have the 2-list organizing structure: one or more helper functions that help keep the main function shorter and reveal the structure of the algorithm.

Reply:
2-list structure enabled a great helper function.